src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,147 @@
+            this.hash = hash;
+            this.key = key;
+            this.value = value;
+            this.next = next;
+        }
+    }
 
     @Override
     public void put(K key, V value) {
+        if (key == null) {
+            putNullKey(value);
+            return;
+        }
+        int hash = hash(key);
+        Node<K, V> newNode = new Node<>(hash, key, value, null);
+        if (table[hash] == null) {
+            table[hash] = newNode;
+        } else {
+            Node<K, V> current = table[hash];
+            while (current != null) {
+                if (current.key != null && current.key.equals(key)) {
+                    current.value = value;
+                    return;
+                }
+                current = current.next;
+            }
+            current = table[hash];
+            while (current.next != null) {
+                current = current.next;
+            }
+            current.next = newNode;
+        }
+        size++;
+        if ((double) size / table.length > LOAD_FACTOR) {
+            resize();
+        }
+    }
+
+    private void putNullKey(V value) {
+        int hash = 0;
+        if (table[hash] == null) {
+            table[hash] = new Node<>(hash, null, value, null);
+        } else {
+            Node<K, V> current = table[hash];
+            while (current != null) {
+                if (current.key == null) {
+                    current.value = value;
+                    return;
+                }
+                current = current.next;
+            }
+            Node<K, V> newNode = new Node<>(hash, null, value, null);
+            newNode.next = table[hash];
+            table[hash] = newNode;
+        }
+        size++;
+        if ((double) size / table.length > LOAD_FACTOR) {
+            resize();
+        }
+    }
 
+    private void resize() {
+        Node<K, V>[] oldTable = table;
+        int newCapacity = table.length * 2;
+        table = new Node[newCapacity];
+        for (Node<K, V> oldNode : oldTable) {
+            while (oldNode != null) {
+                int hash = newHash(oldNode.key, newCapacity);
+                Node<K, V> newNode = new Node<>(hash, oldNode.key, oldNode.value, null);
+                if (table[hash] == null) {
+                    table[hash] = newNode;
+                } else {
+                    Node<K, V> current = table[hash];
+                    while (current.next != null) {
+                        current = current.next;
+                    }
+                    current.next = newNode;
+                }
+                oldNode = oldNode.next;
+            }
+        }
+    }
+
+    private int newHash(K key, int newCapacity) {
+        return (key == null) ? 0 : Math.abs(key.hashCode() % newCapacity);
     }
 
     @Override
     public V getValue(K key) {
+        if (key == null) {
+            return valueForNullKey();
+        }
+        int hash = hash(key);
+        return valueForKey(key, hash);
```

Comment:
`#valueForNullKey` and `#valueForKey` methods have code duplication and such code extraction is redundant. 

These methods should be merged